src/CQRS/LeanCode.CQRS.AspNetCore/Registration/CQRSObjectCollection.cs

```diff
@@ -0,0 +1,97 @@
+using System.Collections.Frozen;
+using System.Reflection;
+using LeanCode.Components;
+using LeanCode.Contracts;
+using LeanCode.CQRS.Execution;
+
+namespace LeanCode.CQRS.AspNetCore.Registration;
+
+public class CQRSObjectCollection : ICQRSObjectSource
+{
+    private readonly HashSet<CQRSObjectMetadata> objects = new(new CQRSObjectMetadataEqualityComparer());
+
+    protected Lazy<FrozenDictionary<Type, CQRSObjectMetadata>> CachedMetadata { get; }
+
+    public IReadOnlySet<CQRSObjectMetadata> Objects => objects;
+
+    public CQRSObjectCollection()
+    {
+        CachedMetadata = new(() => objects.ToFrozenDictionary(m => m.ObjectType), LazyThreadSafetyMode.PublicationOnly);
+    }
+
+    public CQRSObjectMetadata MetadataFor(Type type) => CachedMetadata.Value[type];
+
+    public virtual void AddCQRSObject(CQRSObjectMetadata metadata)
+    {
+        if (CachedMetadata.IsValueCreated)
+        {
+            throw new InvalidOperationException("Cannot add another CQRS object after the source has been frozen.");
+        }
+
+        if (!objects.Add(metadata))
+        {
+            throw new InvalidOperationException(
+                $"CQRS Object({metadata.ObjectKind}) {metadata.ObjectType} is already registered."
+            );
+        }
+    }
+
+    public static IEnumerable<(TypeInfo Contract, CQRSObjectKind ObjectKind)> EnumerateCQRSObjects(
+        TypesCatalog contractsCatalog
+    )
+    {
+        var contracts = contractsCatalog
+            .Assemblies.SelectMany(a => a.DefinedTypes)
+            .Where(t => IsCommand(t) || IsQuery(t) || IsOperation(t));
+
+        foreach (var contract in contracts)
+        {
+            if (ValidateContractType(contract) is not { } objectKind)
+            {
+                continue;
+            }
+
+            yield return (contract, objectKind);
```

Comment:
```suggestion
            if (ValidateContractType(contract) is { } objectKind)
            {
                 yield return (contract, objectKind);
            }
```
This is both shorter and simpler.